bridges/modules/xcm-bridge-hub/src/exporter.rs

```diff
@@ -75,43 +125,215 @@ where
 		);
 
-		// notify XCM queue manager about updated lane state
-		LocalXcmQueueManager::<T::LanesSupport>::on_bridge_message_enqueued(
-			&sender_and_lane,
-			artifacts.enqueued_messages,
-		);
+		// maybe we need switch to congested state
+		Self::on_bridge_message_enqueued(bridge_id, bridge, artifacts.enqueued_messages);
 
 		Ok(id)
 	}
 }
 
+impl<T: Config<I>, I: 'static> OnMessagesDelivered for Pallet<T, I> {
+	fn on_messages_delivered(lane_id: LaneId, enqueued_messages: MessageNonce) {
+		Self::on_bridge_messages_delivered(lane_id, enqueued_messages);
+	}
+}
+
+impl<T: Config<I>, I: 'static> Pallet<T, I> {
+	/// Called when new message is pushed onto outbound bridge queue.
+	fn on_bridge_message_enqueued(
+		bridge_id: BridgeId,
+		bridge: BridgeOf<T, I>,
+		enqueued_messages: MessageNonce,
+	) {
+		// if the bridge queue is not congested, we don't want to do anything
+		let is_congested = enqueued_messages > OUTBOUND_LANE_CONGESTED_THRESHOLD;
+		if !is_congested {
+			return
+		}
+
+		// TODO: https://github.com/paritytech/parity-bridges-common/issues/2006 we either need fishermens
+		// to watch this rule violation (suspended, but keep sending new messages), or we need a
+		// hard limit for that like other XCM queues have
+
+		// check if the lane is already suspended. If it is, do nothing. We still accept new
+		// messages to the suspended bridge, hoping that it'll be actually suspended soon
+		if bridge.state == BridgeState::Suspended {
+			return
+		}
+
+		// else - suspend the bridge
+		let bridge_origin_relative_location = match bridge.bridge_origin_relative_location.try_as()
+		{
+			Ok(bridge_origin_relative_location) => bridge_origin_relative_location,
+			Err(_) => {
+				log::debug!(
+					target: LOG_TARGET,
+					"Failed to convert the bridge {:?} origin location {:?}",
+					bridge_id,
+					bridge.bridge_origin_relative_location,
+				);
+
+				return
+			},
+		};
+		let suspend_result =
+			T::LocalXcmChannelManager::suspend_bridge(bridge_origin_relative_location, bridge_id);
+		match suspend_result {
+			Ok(_) => {
+				log::debug!(
+					target: LOG_TARGET,
+					"Suspended the bridge {:?}, originated by the {:?}",
+					bridge_id,
+					bridge.bridge_origin_relative_location,
+				);
+			},
+			Err(e) => {
+				log::debug!(
+					target: LOG_TARGET,
+					"Failed to suspended the bridge {:?}, originated by the {:?}: {:?}",
+					bridge_id,
+					bridge.bridge_origin_relative_location,
+					e,
+				);
+
+				return
+			},
+		}
+
+		// and remember that we have suspended the bridge
+		Bridges::<T, I>::mutate_extant(bridge_id, |bridge| {
+			bridge.state = BridgeState::Suspended;
+		});
+	}
+
+	/// Must be called whenever we receive a message delivery confirmation.
+	fn on_bridge_messages_delivered(lane_id: LaneId, enqueued_messages: MessageNonce) {
+		// if the bridge queue is still congested, we don't want to do anything
+		let is_congested = enqueued_messages > OUTBOUND_LANE_UNCONGESTED_THRESHOLD;
```

Comment:
Should this be `OUTBOUND_LANE_CONGESTED_THRESHOLD` instead? Since the variable is called `is_congested`

Reply:
Do `enqueued_messages` decrease when the message is delivered? Giving this a chance of being resumed